library/src/main/java/com/google/android/exoplayer/text/eia608/Eia608TrackRenderer.java

```diff
@@ -58,16 +67,40 @@
   private final Handler textRendererHandler;
   private final MediaFormatHolder formatHolder;
   private final SampleHolder sampleHolder;
-  private final StringBuilder captionStringBuilder;
+  private final SpannableStringBuilder captionStringBuilder;
   private final TreeSet<ClosedCaptionList> pendingCaptionLists;
 
   private boolean inputStreamEnded;
   private int captionMode;
   private int captionRowCount;
-  private String caption;
-  private String lastRenderedCaption;
+  private SpannableStringBuilder caption;
+  private SpannableStringBuilder lastRenderedCaption;
   private ClosedCaptionCtrl repeatableControl;
 
+  private static final String TAG = "Eia608Captions";
+
+  // Incoming styling commands are ordered, a higher priority command clears any previous lower
+  // priority commands. Similarly, output SPANs cannot partially overlap, so we need to close all
+  // lower priority SPANs in case of opening a higher priority one. Having these preconditions we
+  // do not need stack or more complex tracking of styles
+  private static final int STYLE_PRIORITY_UNDERLINE = 1; // lowest priority
+  private static final int STYLE_PRIORITY_ITALIC = 2;
+  private static final int STYLE_PRIORITY_COLOR = 3;
+
+  // defined only to give a name to the highest priority (means "every spans").
+  private static final int STYLE_PRIORITY_ALL = 4;
+
+  // Characters (text to display) and control codes (formatting instructions) are arriving
+  // one-by-one, and we want to translate this incoming byte stream into a character sequence with
+  // spans. We need the start and the end position to correctly set up a span, so we will use the
+  // following variables to store information about previously received formatting instructions.
+  // When we get to a point where we should end or change formatting, we can set up the spans
+  // between the stored start position and the current position (new control code, end of line, etc)
+  private int styleSwitchUnderlineStartPos;
```

Comment:
These need to be inited to -1, default 0 means that start this Tag from pos 0 incorrectly.